### ISSUE TITLE
Dynamic (automatic) Slide Counting

### ISSUE THREAD

**Issue by amagura** [codythegreat/dss#7]
## Description
Currently, if a user has/wants fewer or greater than 15 slides, they have to supply a number with the `-s` option.  This should be automatic: the program should auto-detect how many slides there are in a presentation and adjust accordingly.

## Proposed Solution
A function that quickly scans the presentation file, in a memory efficient manner, and returns the number of slides would do nicely.

**Comment by amagura** [comment]
Would we want to change the behavior of the `-s` option so it would cause parsing to halt once slide number `x` is reached or just get rid of the `-s` option?  Is there a reasonable use-case for this?  (It would be `--stop` and not `--slides`).

**Comment by codythegreat** [comment]
@amagura Have you tried DSS with the commits I made last night? The number of slides should be automatically pulled from `parser.c`. The -s flag is not really useful anymore since `parser.c` handles this. The only issue is that the solution to scan the input file isn't what I would call "efficient" as it scans until the start of a slide, and then uses `rewind(FILE)` to go back to the start and begin building the slide structs. 

All of these changes are present in the `parser.c` file.

**Comment by amagura** [comment]
>Have you tried DSS with the commits I made last night? The number of slides should be automatically pulled from `parser.c`. 

I haven't: I've been setting up Linux on my laptop this week, so I haven't been able to test anything.  I was just looking at the code in `main.c`.

> The -s flag is not really useful anymore since `parser.c` handles this. The only issue is that the solution to scan the input file isn't what I would call "efficient" as it scans until the start of a slide, and then uses `rewind(FILE)` to go back to the start and begin building the slide structs.

Well, that's something I could take a look at this weekend then.

**Comment by amagura** [comment]
I'll update this issue later, changing the goal from adding support for automatic slide counting to more efficient slide counting.

**Comment by codythegreat** [comment]
@amagura sounds good. I haven't touched the `x`, `y`, or `slideCount` variables in `main.c` as I saw that you were working on the options parsing and I didn't want to alter something that you were working on.

As a side note with the current implementation of `parser.c` we shouldn't need `x` or `y` either. I currently have the `Slide` struct set to hold slides up to 5000 characters long. I don't like this solution as it will consume a lot of memory for small slides (like in `sample.txt`) but it wouldn't properly hold large slides >5000. I'd like to see if this could be changed into a pointer to a variable length character array. For some reason I couldn't get this to work yesterday. Hopefully we can fix this soon and `areaX` / `areaY` will actually be useful in the sample files.

**Comment by amagura** [comment]
> As a side note with the current implementation of parser.c we shouldn't need x or y either.

Nice.

> I currently have the Slide struct set to hold slides up to 5000 characters long. I don't like this solution as it will consume a lot of memory for small slides (like in sample.txt) but it wouldn't properly hold large slides >5000. I'd like to see if this could be changed into a pointer to a variable length character array.

That's what I'd do.  We're already scanning the presentation to determine how many slides, might as well determine the biggest slide while we're at it.  

I'll look into this when I work on optimizing the scanner.

>  For some reason I couldn't get this to work yesterday. Hopefully we can fix this soon and areaX / areaY will actually be useful in the sample files.

Is the code still there?  I'd like to take a look at it, if possible--just out of curiosity.

**Comment by codythegreat** [comment]
> Is the code still there? I'd like to take a look at it, if possible--just out of curiosity.

@amagura Most of the logic for determining slide size is still present. Pulling of `areaX` and `areaY` values can be found from lines 30-40ish of `parser.c`

as far as I can tell all this would need to work would be the following:
```
int i = 0;
slideContent[slideCount][(x+1)*(y+1)]; // holds each slide array
slideContent[i][0] = '\0'; // prevents junk text by adding line terminator
// ... concatenate to slideContent from file buffer
// ... when {ENDSLIDE} is detected:
slides[i].content = slideContent[i];
```
This is a bit of a simplification. This would go at the tail end of the `parseTXT` function. Only other change I can think of would be changing the struct's content variable to be a pointer.

**Comment by codythegreat** [comment]
Quick update on this: I realized that the loop for finding `areaX`, `areaY`, `title`, and `slides` was actually looping through all the slides because there wasn't a `{STARTSLIDES}` line in the txt files. I added that and remove the call to `rewind` and now the file parsing completes in only one run instead of two. Much more efficient. 

This change is reflected in commit e026c70

**Comment by amagura** [comment]
I'm still fairly unfamiliar with the code base right now, but do you have to draw the slides yourself in the presentation file or do you just need to have `SLIDESTART` and `SLIDEEND` tags before and after each section that should be a slide?

**Comment by codythegreat** [comment]
`STARTSLIDES` goes right before the first slide (only used once in the .txt file as a way to tell the program that all of the meta variables (x, y, s, title) definitions are done.) `ENDSLIDE` goes after each slide to tell the program that a slide is finished being parsed and should be added to the array of `Slide` structs.

If these tags are used properly the program could be set to just grab everything between them as a slide. currently though I have it set to look for a box-like structure consisting of `_` and `|` characters.

**Comment by amagura** [comment]
> STARTSLIDES goes right before the first slide (only used once in the .txt file as a way to tell the program that all of the meta variables (x, y, s, title) definitions are done.)

I'd like to completely remove the need for setting the `x`, `y`, and `s` variables within the presentation.  I think the scanner should determine the `x` and `y` as well as the `s`--hopefully, things I'll be able to accomplish this weekend, time permitting.

> ENDSLIDE goes after each slide to tell the program that a slide is finished being parsed and should be added to the array of Slide structs.

I think it might be nice to have a more settle ending tag, something like `---` or something like that.  

> currently though I have it set to look for a box-like structure consisting of _ and | characters.

Shouldn't the program supply the actual slides and just provide users with a rudimentary (or otherwise) markup language that they can use to format the content on each slide?  Perhaps that's too ambitious *right now*, but I think that should be the parser's end goal.  We could perhaps take a look into using a markdown parser or something like that.<sup>We could even have different styled slides user's could pick from using options.</sup>

**Comment by codythegreat** [comment]
> I'd like to completely remove the need for setting the x, y, and s variables within the presentation. I think the scanner should determine the x and y as well as the s--hopefully, things I'll be able to accomplish this weekend, time permitting.

Yeah the x/y/s values are due to be removed. in both of the sample txt files the `parseTXT` function is pulling the variables set within the files.

> I think it might be nice to have a more settle ending tag, something like --- or something like that.

This would be nice. Currently the parser searches for text patterns within each line of the txt file. I'm using `{ENDSLIDE}` as I think the chances of it ever being used in a presentation is low, but a set of characters like `---` are very likely to be used. If you can find an alternative solution that would allow `---` to reliably be used in a slide without it triggering the end slide logic I'd be for it. 

> Shouldn't the program supply the actual slides and just provide users with a rudimentary (or otherwise) markup language that they can use to format the content on each slide? Perhaps that's too ambitious right now, but I think that should be the parser's end goal. We could perhaps take a look into using a markdown parser or something like that.We could even have different styled slides user's could pick from using options.

One of the main goals is for "raw" formatting (for things like text/ascii art.) I would like to add some light text markup features that would be optional - maybe with flag that could enable/disable them.  I currently have a few changes on my computer that allow for slides to not use any sort of text box structure and pull all of the lines between each `{ENDSLIDE}` tag. I should be able to push this pretty soon, and that will allow for more customization. As for any additional markup/formatting, perhaps we could both collaborate together on how that feature should look and function.

**Comment by codythegreat** [comment]
@amagura Wanted to give you a heads up that I pushed some pretty big commits this morning that might effect what you are doing with this issue. Here is a quick synopsis of the changes:

+ separated struct logic from parser. struct logic is now in `slides.c` and `slides.h`
+ added `struct line` which acts as a linked list where one line leads to the next and so on. This should make it much easier to add per-line formatting as they can just be variables added to the line struct
+ `slide` struct has variable `struct line *first` that points to first line, then each line will point to the next till the last line
+ `parseTXT` function properly allocates data to this new format. `nextLine()` is called after each line to write the buffer to a new line that'll serve as the `*next` variable to the previous line. `parseTXT` will increment to the next `slide` variable and work from that variable's `*first` line variable
+ `displayLoop` now conforms to this new format as well, printing each line of the slide to be shown.
+ after `displayLoop` returns, main will simply free each line and slide, and close.

**Comment by amagura** [comment]
Would it be alright if we don't use typedefs on the structs?  It abstracts the structs without adding any abstractions.

**Comment by codythegreat** [comment]
If you think it'd help in readability or some other way go for. There are a lot of areas that you'll have to change definitions/finction calls. Might be able to find most of them with grep if you are able to use that.

I have a lot of issues that I'll be posting later (just ideas I've thought of) and I'm going to work on cleaning up the code some more tonight before adding anything else.

**Comment by amagura** [comment]
> If you think it'd help in readability or some other way go for. 

Yeah, I think it makes it less confusing to outsiders when reading through code if we don't abstract things unnecessarily.

> There are a lot of areas that you'll have to change definitions/function calls. Might be able to find most of them with grep if you are able to use that.

My text editor flags function definitions/calls when there are inconsistencies between files, so it's pretty easy to find things once you've changed one of them.  And I've actually already started the process, but I got caught up in refactoring things. 

> I have a lot of issues that I'll be posting later (just ideas I've thought of) and I'm going to work on cleaning up the code some more tonight before adding anything else.

That sounds great.

**Comment by amagura** [comment]
Would it be acceptable to assume if one slide has certain dimensions that all the other slides share those dimensions or should we check every slide for different dimensions?

**Comment by codythegreat** [comment]
@amagura I think that it is important that the parser checks each slide for dimensions. Mainly due to the slides being rendered in the center of the screen (which the `x` and `y` variables are used for.)

Since the parser is now scanning slides between each `{ENDSLIDE}` tag this means that slides can take on any form or dimension. I have a new sample slide txt file that I've been working on to showcase how this can be useful.

**Comment by amagura** [comment]
> I think that it is important that the parser checks each slide for dimensions. Mainly due to the slides being rendered in the center of the screen (which the x and y variables are used for.)

Alright, I'll need to change some of the code I've written then.  

> Since the parser is now scanning slides between each {ENDSLIDE} tag this means that slides can take on any form or dimension. I have a new sample slide txt file that I've been working on to showcase how this can be useful.

I think my `scanr` function looks for `.start{}` and `.end{}` and uses those to delineate slides--it was a pretty arbitrary decision to use those strings, so it's not a big deal if you want to change them.  I think it might be nice to have a unambiguous meta data section in the slide file.  Something like a JSON object or something similar that details information like Author, Title, Date, other options perhaps?

**Comment by codythegreat** [comment]
@amagura The meta data idea is great! I like the idea of a JSON format as most people are at least familiar with it, and I'm sure it'd be much easier to parse data in an industry standard format.

Those tags for start and end are perfectly fine. 

It'd be nice if there were some way we could get the slide count from a file without having to parse over it twice and without the user having to explicitly define it.

**Comment by amagura** [comment]
> It'd be nice if there were some way we could get the slide count from a file without having to parse over it twice and without the user having to explicitly define it.

I don't think we can have both those things: if we want the program to automatically count how many slides are present then it needs to read the file before the main program starts the presentation.  I don't think this is an issue, however, since rewinding the file pointer does not noticeably impact performance.

**Comment by codythegreat** [comment]
@amagura Good point. We should work that feature in so that the user won't have to continually keep track of slide counts especially on large projects.

**Comment by amagura** [comment]
I think it may already be a feature in my master branch; if it isn't, it won't be hard to add.
